### PR TITLE
Liquids refactor

### LINKED ISSUE #133: Flow calculations need a total rewrite
### Issue description
We recently fixed a CPU leak relating to still water bodies. Flow calculations are still TERRIBLE for performance and need rewriting.

### OS and versions

* PocketMine-MP: 756760a
* PHP: 7.0.13 x64
* Server OS: Linux
* Game version: irrelevant

### Crashdump, backtrace or other files

* ...

### PR DESCRIPTION
**Pull request by dktapps**
- [x] Performance improvements
- [x] Fixed liquid going crazy when flowing downhill
- [x] Fixed CPU leaks from still liquid flowing into other still liquid
- [x] Fixed lava being impossible to get rid of properly

### Relevant issues
fixes #133 

## TODO
- more performance improvements
- implement static and dynamic liquids properly

### CODE REVIEW

File: src/pocketmine/block/Lava.php [hunk 1]
```diff
@@ -57,23 +57,27 @@ public function getFlowDecayPerBlock() : int{
 	}
 
 	protected function checkForHarden(){
-		$colliding = false;
-		for($side = 1; $side <= 5 and !$colliding; ++$side){ //don't check downwards side
-			$colliding = $this->getSide($side) instanceof Water;
+		$colliding = null;
+		for($side = 1; $side <= 5; ++$side){ //don't check downwards side
+			$blockSide = $this->getSide($side);
+			if($side instanceof Water){
```
Comment (zKoz210) [inline src/pocketmine/block/Lava.php]: BlockSide?

Comment (dktapps) [reply src/pocketmine/block/Lava.php]: hmm, wtf

**Comment by zKoz210** [comment]
Need particle lava/water mix?

**Comment by dktapps** [comment]
@zKoz210 I'm not sure about that. Vanilla doesn't have particles at the moment either. I think they are supposed to be client-sided.

**Comment by dktapps** [comment]
I'm going to merge this as-is since it brings a lot of much-needed bugfixes and performance improvements. While implementing still liquids correctly is desirable, it is not necessary for liquids to function sanely (for now).

**Pull request merged**